### ISSUE TITLE
SFTP: Connection closed prematurely

### ISSUE THREAD

**Issue by jagermesh** [phpseclib/phpseclib#1741]
I'm having problem with SFTP connection via login/password (not the key) after upgrading to version 3. It was working fine in version 2.

```PHP
$sftp= new SFTP($hostName, $port);
$sftp->login($userName, $password);
```

Call stack

```
phpseclib3\Net\SSH2->send_binary_packet('.......');
    in vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php, 1724
phpseclib3\Net\SSH2->key_exchange();
    in vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php, 1420
phpseclib3\Net\SSH2->connect();
    in vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php, 2167
phpseclib3\Net\SSH2->sublogin('XXXXXX');
    in vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php, 2145
```

Logs

```
00000000  53:53:48:2d:32:2e:30:2d:4f:70:65:6e:53:53:48:5f  SSH-2.0-OpenSSH_
00000010  36:2e:32:0d:0a                                   6.2..

->
00000000  53:53:48:2d:32:2e:30:2d:70:68:70:73:65:63:6c:69  SSH-2.0-phpsecli
00000010  62:5f:33:2e:30:20:28:6c:69:62:73:6f:64:69:75:6d  b_3.0 (libsodium
00000020  2c:20:6f:70:65:6e:73:73:6c:2c:20:67:6d:70:29:0d  , openssl, gmp).
00000030  0a                                               .

-> NET_SSH2_MSG_KEXINIT (since last: 0.1268, network: 0s)
00000000  76:d0:bb:ef:7c:12:26:28:74:e3:71:3a:97:e9:bc:97  v...|.&(t.q:....
00000010  00:00:01:7c:63:75:72:76:65:32:35:35:31:39:2d:73  ...|curve25519-s
00000020  68:61:32:35:36:2c:63:75:72:76:65:32:35:35:31:39  ha256,curve25519
00000030  2d:73:68:61:32:35:36:40:6c:69:62:73:73:68:2e:6f  -sha256@libssh.o
00000040  72:67:2c:65:63:64:68:2d:73:68:61:32:2d:6e:69:73  rg,ecdh-sha2-nis
00000050  74:70:32:35:36:2c:65:63:64:68:2d:73:68:61:32:2d  tp256,ecdh-sha2-
00000060  6e:69:73:74:70:33:38:34:2c:65:63:64:68:2d:73:68  nistp384,ecdh-sh
00000070  61:32:2d:6e:69:73:74:70:35:32:31:2c:64:69:66:66  a2-nistp521,diff
00000080  69:65:2d:68:65:6c:6c:6d:61:6e:2d:67:72:6f:75:70  ie-hellman-group
00000090  2d:65:78:63:68:61:6e:67:65:2d:73:68:61:32:35:36  -exchange-sha256
000000a0  2c:64:69:66:66:69:65:2d:68:65:6c:6c:6d:61:6e:2d  ,diffie-hellman-
000000b0  67:72:6f:75:70:2d:65:78:63:68:61:6e:67:65:2d:73  group-exchange-s
000000c0  68:61:31:2c:64:69:66:66:69:65:2d:68:65:6c:6c:6d  ha1,diffie-hellm
000000d0  61:6e:2d:67:72:6f:75:70:31:34:2d:73:68:61:32:35  an-group14-sha25
000000e0  36:2c:64:69:66:66:69:65:2d:68:65:6c:6c:6d:61:6e  6,diffie-hellman
000000f0  2d:67:72:6f:75:70:31:34:2d:73:68:61:31:2c:64:69  -group14-sha1,di
00000100  66:66:69:65:2d:68:65:6c:6c:6d:61:6e:2d:67:72:6f  ffie-hellman-gro
00000110  75:70:31:35:2d:73:68:61:35:31:32:2c:64:69:66:66  up15-sha512,diff
00000120  69:65:2d:68:65:6c:6c:6d:61:6e:2d:67:72:6f:75:70  ie-hellman-group
00000130  31:36:2d:73:68:61:35:31:32:2c:64:69:66:66:69:65  16-sha512,diffie
00000140  2d:68:65:6c:6c:6d:61:6e:2d:67:72:6f:75:70:31:37  -hellman-group17
00000150  2d:73:68:61:35:31:32:2c:64:69:66:66:69:65:2d:68  -sha512,diffie-h
00000160  65:6c:6c:6d:61:6e:2d:67:72:6f:75:70:31:38:2d:73  ellman-group18-s
00000170  68:61:35:31:32:2c:64:69:66:66:69:65:2d:68:65:6c  ha512,diffie-hel
00000180  6c:6d:61:6e:2d:67:72:6f:75:70:31:2d:73:68:61:31  lman-group1-sha1
00000190  00:00:00:71:73:73:68:2d:65:64:32:35:35:31:39:2c  ...qssh-ed25519,
000001a0  65:63:64:73:61:2d:73:68:61:32:2d:6e:69:73:74:70  ecdsa-sha2-nistp
000001b0  32:35:36:2c:65:63:64:73:61:2d:73:68:61:32:2d:6e  256,ecdsa-sha2-n
000001c0  69:73:74:70:33:38:34:2c:65:63:64:73:61:2d:73:68  istp384,ecdsa-sh
000001d0  61:32:2d:6e:69:73:74:70:35:32:31:2c:72:73:61:2d  a2-nistp521,rsa-
000001e0  73:68:61:32:2d:32:35:36:2c:72:73:61:2d:73:68:61  sha2-256,rsa-sha
000001f0  32:2d:35:31:32:2c:73:73:68:2d:72:73:61:2c:73:73  2-512,ssh-rsa,ss
00000200  68:2d:64:73:73:00:00:01:01:61:65:73:31:32:38:2d  h-dss....aes128-
00000210  67:63:6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c  gcm@openssh.com,
00000220  61:65:73:32:35:36:2d:67:63:6d:40:6f:70:65:6e:73  aes256-gcm@opens
00000230  73:68:2e:63:6f:6d:2c:61:65:73:31:32:38:2d:63:74  sh.com,aes128-ct
00000240  72:2c:61:65:73:31:39:32:2d:63:74:72:2c:61:65:73  r,aes192-ctr,aes
00000250  32:35:36:2d:63:74:72:2c:61:65:73:31:32:38:2d:63  256-ctr,aes128-c
00000260  62:63:2c:61:65:73:31:39:32:2d:63:62:63:2c:61:65  bc,aes192-cbc,ae
00000270  73:32:35:36:2d:63:62:63:2c:62:6c:6f:77:66:69:73  s256-cbc,blowfis
00000280  68:2d:63:74:72:2c:62:6c:6f:77:66:69:73:68:2d:63  h-ctr,blowfish-c
00000290  62:63:2c:33:64:65:73:2d:63:74:72:2c:33:64:65:73  bc,3des-ctr,3des
000002a0  2d:63:62:63:2c:74:77:6f:66:69:73:68:31:32:38:2d  -cbc,twofish128-
000002b0  63:74:72:2c:74:77:6f:66:69:73:68:31:39:32:2d:63  ctr,twofish192-c
000002c0  74:72:2c:74:77:6f:66:69:73:68:32:35:36:2d:63:74  tr,twofish256-ct
000002d0  72:2c:74:77:6f:66:69:73:68:31:32:38:2d:63:62:63  r,twofish128-cbc
000002e0  2c:74:77:6f:66:69:73:68:31:39:32:2d:63:62:63:2c  ,twofish192-cbc,
000002f0  74:77:6f:66:69:73:68:32:35:36:2d:63:62:63:2c:74  twofish256-cbc,t
00000300  77:6f:66:69:73:68:2d:63:62:63:00:00:01:01:61:65  wofish-cbc....ae
00000310  73:31:32:38:2d:67:63:6d:40:6f:70:65:6e:73:73:68  s128-gcm@openssh
00000320  2e:63:6f:6d:2c:61:65:73:32:35:36:2d:67:63:6d:40  .com,aes256-gcm@
00000330  6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:61:65:73:31  openssh.com,aes1
00000340  32:38:2d:63:74:72:2c:61:65:73:31:39:32:2d:63:74  28-ctr,aes192-ct
00000350  72:2c:61:65:73:32:35:36:2d:63:74:72:2c:61:65:73  r,aes256-ctr,aes
00000360  31:32:38:2d:63:62:63:2c:61:65:73:31:39:32:2d:63  128-cbc,aes192-c
00000370  62:63:2c:61:65:73:32:35:36:2d:63:62:63:2c:62:6c  bc,aes256-cbc,bl
00000380  6f:77:66:69:73:68:2d:63:74:72:2c:62:6c:6f:77:66  owfish-ctr,blowf
00000390  69:73:68:2d:63:62:63:2c:33:64:65:73:2d:63:74:72  ish-cbc,3des-ctr
000003a0  2c:33:64:65:73:2d:63:62:63:2c:74:77:6f:66:69:73  ,3des-cbc,twofis
000003b0  68:31:32:38:2d:63:74:72:2c:74:77:6f:66:69:73:68  h128-ctr,twofish
000003c0  31:39:32:2d:63:74:72:2c:74:77:6f:66:69:73:68:32  192-ctr,twofish2
000003d0  35:36:2d:63:74:72:2c:74:77:6f:66:69:73:68:31:32  56-ctr,twofish12
000003e0  38:2d:63:62:63:2c:74:77:6f:66:69:73:68:31:39:32  8-cbc,twofish192
000003f0  2d:63:62:63:2c:74:77:6f:66:69:73:68:32:35:36:2d  -cbc,twofish256-
00000400  63:62:63:2c:74:77:6f:66:69:73:68:2d:63:62:63:00  cbc,twofish-cbc.
00000410  00:00:f7:68:6d:61:63:2d:73:68:61:32:2d:32:35:36  ...hmac-sha2-256
00000420  2d:65:74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d  -etm@openssh.com
00000430  2c:68:6d:61:63:2d:73:68:61:32:2d:35:31:32:2d:65  ,hmac-sha2-512-e
00000440  74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:75  tm@openssh.com,u
00000450  6d:61:63:2d:36:34:2d:65:74:6d:40:6f:70:65:6e:73  mac-64-etm@opens
00000460  73:68:2e:63:6f:6d:2c:75:6d:61:63:2d:31:32:38:2d  sh.com,umac-128-
00000470  65:74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c  etm@openssh.com,
00000480  68:6d:61:63:2d:73:68:61:31:2d:65:74:6d:40:6f:70  hmac-sha1-etm@op
00000490  65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d:73  enssh.com,hmac-s
000004a0  68:61:32:2d:32:35:36:2c:68:6d:61:63:2d:73:68:61  ha2-256,hmac-sha
000004b0  32:2d:35:31:32:2c:75:6d:61:63:2d:36:34:40:6f:70  2-512,umac-64@op
000004c0  65:6e:73:73:68:2e:63:6f:6d:2c:75:6d:61:63:2d:31  enssh.com,umac-1
000004d0  32:38:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68  28@openssh.com,h
000004e0  6d:61:63:2d:73:68:61:31:2d:39:36:2c:68:6d:61:63  mac-sha1-96,hmac
000004f0  2d:73:68:61:31:2c:68:6d:61:63:2d:6d:64:35:2d:39  -sha1,hmac-md5-9
00000500  36:2c:68:6d:61:63:2d:6d:64:35:00:00:00:f7:68:6d  6,hmac-md5....hm
00000510  61:63:2d:73:68:61:32:2d:32:35:36:2d:65:74:6d:40  ac-sha2-256-etm@
00000520  6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63  openssh.com,hmac
00000530  2d:73:68:61:32:2d:35:31:32:2d:65:74:6d:40:6f:70  -sha2-512-etm@op
00000540  65:6e:73:73:68:2e:63:6f:6d:2c:75:6d:61:63:2d:36  enssh.com,umac-6
00000550  34:2d:65:74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f  4-etm@openssh.co
00000560  6d:2c:75:6d:61:63:2d:31:32:38:2d:65:74:6d:40:6f  m,umac-128-etm@o
00000570  70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d  penssh.com,hmac-
00000580  73:68:61:31:2d:65:74:6d:40:6f:70:65:6e:73:73:68  sha1-etm@openssh
00000590  2e:63:6f:6d:2c:68:6d:61:63:2d:73:68:61:32:2d:32  .com,hmac-sha2-2
000005a0  35:36:2c:68:6d:61:63:2d:73:68:61:32:2d:35:31:32  56,hmac-sha2-512
000005b0  2c:75:6d:61:63:2d:36:34:40:6f:70:65:6e:73:73:68  ,umac-64@openssh
000005c0  2e:63:6f:6d:2c:75:6d:61:63:2d:31:32:38:40:6f:70  .com,umac-128@op
000005d0  65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d:73  enssh.com,hmac-s
000005e0  68:61:31:2d:39:36:2c:68:6d:61:63:2d:73:68:61:31  ha1-96,hmac-sha1
000005f0  2c:68:6d:61:63:2d:6d:64:35:2d:39:36:2c:68:6d:61  ,hmac-md5-96,hma
00000600  63:2d:6d:64:35:00:00:00:1a:6e:6f:6e:65:2c:7a:6c  c-md5....none,zl
00000610  69:62:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:7a  ib@openssh.com,z
00000620  6c:69:62:00:00:00:1a:6e:6f:6e:65:2c:7a:6c:69:62  lib....none,zlib
00000630  40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:7a:6c:69  @openssh.com,zli
00000640  62:00:00:00:00:00:00:00:00:00:00:00:00:00        b.............

<- NET_SSH2_MSG_KEXINIT (since last: 0.0415, network: 0.0415s)
00000000  79:ca:af:fb:dd:2a:17:2d:e9:6b:0e:28:51:64:45:a2  y....*.-.k.(QdE.
00000010  00:00:00:b7:65:63:64:68:2d:73:68:61:32:2d:6e:69  ....ecdh-sha2-ni
00000020  73:74:70:32:35:36:2c:65:63:64:68:2d:73:68:61:32  stp256,ecdh-sha2
00000030  2d:6e:69:73:74:70:33:38:34:2c:65:63:64:68:2d:73  -nistp384,ecdh-s
00000040  68:61:32:2d:6e:69:73:74:70:35:32:31:2c:64:69:66  ha2-nistp521,dif
00000050  66:69:65:2d:68:65:6c:6c:6d:61:6e:2d:67:72:6f:75  fie-hellman-grou
00000060  70:2d:65:78:63:68:61:6e:67:65:2d:73:68:61:32:35  p-exchange-sha25
00000070  36:2c:64:69:66:66:69:65:2d:68:65:6c:6c:6d:61:6e  6,diffie-hellman
00000080  2d:67:72:6f:75:70:2d:65:78:63:68:61:6e:67:65:2d  -group-exchange-
00000090  73:68:61:31:2c:64:69:66:66:69:65:2d:68:65:6c:6c  sha1,diffie-hell
000000a0  6d:61:6e:2d:67:72:6f:75:70:31:34:2d:73:68:61:31  man-group14-sha1
000000b0  2c:64:69:66:66:69:65:2d:68:65:6c:6c:6d:61:6e:2d  ,diffie-hellman-
000000c0  67:72:6f:75:70:31:2d:73:68:61:31:00:00:00:23:73  group1-sha1...#s
000000d0  73:68:2d:72:73:61:2c:73:73:68:2d:64:73:73:2c:65  sh-rsa,ssh-dss,e
000000e0  63:64:73:61:2d:73:68:61:32:2d:6e:69:73:74:70:32  cdsa-sha2-nistp2
000000f0  35:36:00:00:00:cb:61:65:73:31:32:38:2d:63:74:72  56....aes128-ctr
00000100  2c:61:65:73:31:39:32:2d:63:74:72:2c:61:65:73:32  ,aes192-ctr,aes2
00000110  35:36:2d:63:74:72:2c:61:72:63:66:6f:75:72:32:35  56-ctr,arcfour25
00000120  36:2c:61:72:63:66:6f:75:72:31:32:38:2c:61:65:73  6,arcfour128,aes
00000130  31:32:38:2d:67:63:6d:40:6f:70:65:6e:73:73:68:2e  128-gcm@openssh.
00000140  63:6f:6d:2c:61:65:73:32:35:36:2d:67:63:6d:40:6f  com,aes256-gcm@o
00000150  70:65:6e:73:73:68:2e:63:6f:6d:2c:61:65:73:31:32  penssh.com,aes12
00000160  38:2d:63:62:63:2c:33:64:65:73:2d:63:62:63:2c:62  8-cbc,3des-cbc,b
00000170  6c:6f:77:66:69:73:68:2d:63:62:63:2c:63:61:73:74  lowfish-cbc,cast
00000180  31:32:38:2d:63:62:63:2c:61:65:73:31:39:32:2d:63  128-cbc,aes192-c
00000190  62:63:2c:61:65:73:32:35:36:2d:63:62:63:2c:61:72  bc,aes256-cbc,ar
000001a0  63:66:6f:75:72:2c:72:69:6a:6e:64:61:65:6c:2d:63  cfour,rijndael-c
000001b0  62:63:40:6c:79:73:61:74:6f:72:2e:6c:69:75:2e:73  bc@lysator.liu.s
000001c0  65:00:00:00:cb:61:65:73:31:32:38:2d:63:74:72:2c  e....aes128-ctr,
000001d0  61:65:73:31:39:32:2d:63:74:72:2c:61:65:73:32:35  aes192-ctr,aes25
000001e0  36:2d:63:74:72:2c:61:72:63:66:6f:75:72:32:35:36  6-ctr,arcfour256
000001f0  2c:61:72:63:66:6f:75:72:31:32:38:2c:61:65:73:31  ,arcfour128,aes1
00000200  32:38:2d:67:63:6d:40:6f:70:65:6e:73:73:68:2e:63  28-gcm@openssh.c
00000210  6f:6d:2c:61:65:73:32:35:36:2d:67:63:6d:40:6f:70  om,aes256-gcm@op
00000220  65:6e:73:73:68:2e:63:6f:6d:2c:61:65:73:31:32:38  enssh.com,aes128
00000230  2d:63:62:63:2c:33:64:65:73:2d:63:62:63:2c:62:6c  -cbc,3des-cbc,bl
00000240  6f:77:66:69:73:68:2d:63:62:63:2c:63:61:73:74:31  owfish-cbc,cast1
00000250  32:38:2d:63:62:63:2c:61:65:73:31:39:32:2d:63:62  28-cbc,aes192-cb
00000260  63:2c:61:65:73:32:35:36:2d:63:62:63:2c:61:72:63  c,aes256-cbc,arc
00000270  66:6f:75:72:2c:72:69:6a:6e:64:61:65:6c:2d:63:62  four,rijndael-cb
00000280  63:40:6c:79:73:61:74:6f:72:2e:6c:69:75:2e:73:65  c@lysator.liu.se
00000290  00:00:01:92:68:6d:61:63:2d:6d:64:35:2d:65:74:6d  ....hmac-md5-etm
000002a0  40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61  @openssh.com,hma
000002b0  63:2d:73:68:61:31:2d:65:74:6d:40:6f:70:65:6e:73  c-sha1-etm@opens
000002c0  73:68:2e:63:6f:6d:2c:75:6d:61:63:2d:36:34:2d:65  sh.com,umac-64-e
000002d0  74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:75  tm@openssh.com,u
000002e0  6d:61:63:2d:31:32:38:2d:65:74:6d:40:6f:70:65:6e  mac-128-etm@open
000002f0  73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d:73:68:61  ssh.com,hmac-sha
00000300  32:2d:32:35:36:2d:65:74:6d:40:6f:70:65:6e:73:73  2-256-etm@openss
00000310  68:2e:63:6f:6d:2c:68:6d:61:63:2d:73:68:61:32:2d  h.com,hmac-sha2-
00000320  35:31:32:2d:65:74:6d:40:6f:70:65:6e:73:73:68:2e  512-etm@openssh.
00000330  63:6f:6d:2c:68:6d:61:63:2d:72:69:70:65:6d:64:31  com,hmac-ripemd1
00000340  36:30:2d:65:74:6d:40:6f:70:65:6e:73:73:68:2e:63  60-etm@openssh.c
00000350  6f:6d:2c:68:6d:61:63:2d:73:68:61:31:2d:39:36:2d  om,hmac-sha1-96-
00000360  65:74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c  etm@openssh.com,
00000370  68:6d:61:63:2d:6d:64:35:2d:39:36:2d:65:74:6d:40  hmac-md5-96-etm@
00000380  6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63  openssh.com,hmac
00000390  2d:6d:64:35:2c:68:6d:61:63:2d:73:68:61:31:2c:75  -md5,hmac-sha1,u
000003a0  6d:61:63:2d:36:34:40:6f:70:65:6e:73:73:68:2e:63  mac-64@openssh.c
000003b0  6f:6d:2c:75:6d:61:63:2d:31:32:38:40:6f:70:65:6e  om,umac-128@open
000003c0  73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d:73:68:61  ssh.com,hmac-sha
000003d0  32:2d:32:35:36:2c:68:6d:61:63:2d:73:68:61:32:2d  2-256,hmac-sha2-
000003e0  35:31:32:2c:68:6d:61:63:2d:72:69:70:65:6d:64:31  512,hmac-ripemd1
000003f0  36:30:2c:68:6d:61:63:2d:72:69:70:65:6d:64:31:36  60,hmac-ripemd16
00000400  30:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d  0@openssh.com,hm
00000410  61:63:2d:73:68:61:31:2d:39:36:2c:68:6d:61:63:2d  ac-sha1-96,hmac-
00000420  6d:64:35:2d:39:36:00:00:01:92:68:6d:61:63:2d:6d  md5-96....hmac-m
00000430  64:35:2d:65:74:6d:40:6f:70:65:6e:73:73:68:2e:63  d5-etm@openssh.c
00000440  6f:6d:2c:68:6d:61:63:2d:73:68:61:31:2d:65:74:6d  om,hmac-sha1-etm
00000450  40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:75:6d:61  @openssh.com,uma
00000460  63:2d:36:34:2d:65:74:6d:40:6f:70:65:6e:73:73:68  c-64-etm@openssh
00000470  2e:63:6f:6d:2c:75:6d:61:63:2d:31:32:38:2d:65:74  .com,umac-128-et
00000480  6d:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d  m@openssh.com,hm
00000490  61:63:2d:73:68:61:32:2d:32:35:36:2d:65:74:6d:40  ac-sha2-256-etm@
000004a0  6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63  openssh.com,hmac
000004b0  2d:73:68:61:32:2d:35:31:32:2d:65:74:6d:40:6f:70  -sha2-512-etm@op
000004c0  65:6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d:72  enssh.com,hmac-r
000004d0  69:70:65:6d:64:31:36:30:2d:65:74:6d:40:6f:70:65  ipemd160-etm@ope
000004e0  6e:73:73:68:2e:63:6f:6d:2c:68:6d:61:63:2d:73:68  nssh.com,hmac-sh
000004f0  61:31:2d:39:36:2d:65:74:6d:40:6f:70:65:6e:73:73  a1-96-etm@openss
00000500  68:2e:63:6f:6d:2c:68:6d:61:63:2d:6d:64:35:2d:39  h.com,hmac-md5-9
00000510  36:2d:65:74:6d:40:6f:70:65:6e:73:73:68:2e:63:6f  6-etm@openssh.co
00000520  6d:2c:68:6d:61:63:2d:6d:64:35:2c:68:6d:61:63:2d  m,hmac-md5,hmac-
00000530  73:68:61:31:2c:75:6d:61:63:2d:36:34:40:6f:70:65  sha1,umac-64@ope
00000540  6e:73:73:68:2e:63:6f:6d:2c:75:6d:61:63:2d:31:32  nssh.com,umac-12
00000550  38:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:2c:68:6d  8@openssh.com,hm
00000560  61:63:2d:73:68:61:32:2d:32:35:36:2c:68:6d:61:63  ac-sha2-256,hmac
00000570  2d:73:68:61:32:2d:35:31:32:2c:68:6d:61:63:2d:72  -sha2-512,hmac-r
00000580  69:70:65:6d:64:31:36:30:2c:68:6d:61:63:2d:72:69  ipemd160,hmac-ri
00000590  70:65:6d:64:31:36:30:40:6f:70:65:6e:73:73:68:2e  pemd160@openssh.
000005a0  63:6f:6d:2c:68:6d:61:63:2d:73:68:61:31:2d:39:36  com,hmac-sha1-96
000005b0  2c:68:6d:61:63:2d:6d:64:35:2d:39:36:00:00:00:15  ,hmac-md5-96....
000005c0  6e:6f:6e:65:2c:7a:6c:69:62:40:6f:70:65:6e:73:73  none,zlib@openss
000005d0  68:2e:63:6f:6d:00:00:00:15:6e:6f:6e:65:2c:7a:6c  h.com....none,zl
000005e0  69:62:40:6f:70:65:6e:73:73:68:2e:63:6f:6d:00:00  ib@openssh.com..
000005f0  00:00:00:00:00:00:00:00:00:00:00                 ...........
```

Any help will be appreciated...

**Comment by terrafrost** [comment]
It'd be interesting to know what the SSH server is saying. Is the SSH server one you have control over? If so then could you do `sudo /usr/sbin/sshd -ddd` and lmk what the output is?

Nothing in what you've posted looks suspicious: my guess is that it's on the server end.

**Comment by jagermesh** [comment]
Nope, I don't have control and permission. Below is what I can do. Will this help?

```
$ /usr/sbin/sshd -ddd
debug2: load_server_config: filename /etc/ssh/sshd_config
/etc/ssh/sshd_config: Permission denied

$ ssh -v localhost
OpenSSH_6.2p2, OpenSSL 0.9.8j-fips 07 Jan 2009
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 20: Applying options for *
debug1: Connecting to localhost [127.0.0.1] port 22.
debug1: Connection established.
debug1: SELinux support disabled
debug1: identity file /home/......................................./.ssh/id_rsa type -1
debug1: identity file /home/......................................./.ssh/id_rsa-cert type -1
debug1: identity file /home/......................................./.ssh/id_dsa type -1
debug1: identity file /home/......................................./.ssh/id_dsa-cert type -1
debug1: identity file /home/......................................./.ssh/id_ecdsa type -1
debug1: identity file /home/......................................./.ssh/id_ecdsa-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_6.2
debug1: Remote protocol version 2.0, remote software version OpenSSH_6.2
debug1: match: OpenSSH_6.2 pat OpenSSH*
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: server->client aes128-ctr hmac-md5-etm@openssh.com none
debug1: kex: client->server aes128-ctr hmac-md5-etm@openssh.com none
debug1: sending SSH2_MSG_KEX_ECDH_INIT
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ECDSA ....................................... [MD5]
The authenticity of host 'localhost (127.0.0.1)' can't be established.
ECDSA key fingerprint is ....................................... [MD5].
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'localhost' (ECDSA) to the list of known hosts.
debug1: ssh_ecdsa_verify: signature correct
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: Roaming not allowed by server
debug1: SSH2_MSG_SERVICE_REQUEST sent
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Trying private key: /home/......................................./.ssh/id_rsa
debug1: Trying private key: /home/......................................./.ssh/id_dsa
debug1: Trying private key: /home/......................................./.ssh/id_ecdsa
debug1: Next authentication method: keyboard-interactive
Password:
debug1: Authentication succeeded (keyboard-interactive).
Authenticated to localhost ([127.0.0.1]:22).
debug1: channel 0: new [client-session]
debug1: Requesting no-more-sessions@openssh.com
debug1: Entering interactive session.
debug1: Sending environment.
debug1: Sending env LANG = en_US.UTF-8
```

**Comment by terrafrost** [comment]
That's not super helpful to me. I don't want to see how another client is succeeding - I want to see why phpseclib's connection attempt is failing.

phpseclib v3 did introduce support for a bunch of new algorithms. Based on what you've posted phpseclib and the server should be using the following algorithm suite:

```
Array
(
    [kex] => ecdh-sha2-nistp256
    [hostkey] => ecdsa-sha2-nistp256
    [client_to_server] => Array
        (
            [crypt] => aes128-gcm@openssh.com
            [mac] => hmac-sha2-256-etm@openssh.com
            [comp] => none
        )

    [server_to_client] => Array
        (
            [crypt] => aes128-gcm@openssh.com
            [mac] => hmac-sha2-256-etm@openssh.com
            [comp] => none
        )

)
```
The server says it supports those algorithms and thus phpseclib v3 is using them. It's almost like the server actually _doesn't_ support those algorithms and is killing the connection because no compatible algorithms can be found, even tho it _says_ it supports those algorithms.

It's not clear to me what the kex or hostkey algorithms are when you do `ssh -v localhost` but the other algorithms are definitely different:

```
debug1: kex: server->client aes128-ctr hmac-md5-etm@openssh.com none
debug1: kex: client->server aes128-ctr hmac-md5-etm@openssh.com none
```
ie. aes128-ctr instead of aes128-gcm@openssh.com, etc.

You can make phpseclib v3 use those same algorithms by doing this right after the constructor:

```php
$methods = ['crypt' => ['aes128-ctr'], 'mac' => ['hmac-md5-etm@openssh.com']];
$ssh->setPreferredAlgorithms(['client_to_server' => $methods, 'server_to_client' => $methods]);
```
Of course the server could be dying because of the kex and hostkey algorithms too. 

If you wanted to make phpseclib v3 use the same algorithms as v2 you could do this:



```php
$methods = [
    'crypt' => [
        'arcfour256',
        'arcfour128',
        'aes128-ctr',
        'aes192-ctr',
        'aes256-ctr',
        'twofish128-ctr',
        'twofish192-ctr',
        'twofish256-ctr',
        'aes128-cbc',
        'aes192-cbc',
        'aes256-cbc',
        'twofish128-cbc',
        'twofish192-cbc',
        'twofish256-cbc',
        'twofish-cbc',
        'blowfish-ctr',
        'blowfish-cbc',
        '3des-ctr',
        '3des-cbc'
    ],
    'mac' => [
        'hmac-sha2-256',
        'hmac-sha1-96',
        'hmac-sha1',
        'hmac-md5-96',
        'hmac-md5'
    ],
    'comp' => ['none']
];
$ssh->setPreferredAlgorithms([
    'kex' => [
        'curve25519-sha256@libssh.org',
        'diffie-hellman-group-exchange-sha256',
        'diffie-hellman-group-exchange-sha1',
        'diffie-hellman-group14-sha1',
        'diffie-hellman-group14-sha14'
    ],
    'hostkey' => [
        'rsa-sha2-256',
        'rsa-sha2-512',
        'ssh-rsa',
        'ssh-dss'
    ],
    'client_to_server' => $methods,
    'server_to_client' => $methods
]);
```
It actually gets a little more complicated than this because phpseclib rearranges the priority of the symmetric key algorithms based on the extensions that you have installed (it's been doing this since [Sept 7, 2019](https://github.com/phpseclib/phpseclib/commit/3ad44cb3d60c4c403e7a94c80ec04913ba99631c)). I'm not doing that here but whatever.

Anyway, TLDR, that's my best guess. To know for sure I'd need the `sudo /usr/sbin/sshd -ddd`. If I had that I could Google the error message, maybe post on the OpenSSH internals mailing list and look into changing how phpseclib v3 behaves for versions of OpenSSH older than whatever depending on what the OpenSSH people say.

In lieu of that I think the above should work.

**Comment by jagermesh** [comment]
> That's not super helpful to me. I don't want to see how another client is succeeding - I want to see why phpseclib's connection attempt is failing.

I understand. I just tried to give you as much data as as I can, since I don't have sudo permissions on that server.

> In lieu of that I think the above should work.

Thank you! I'll check it out. From around 200+ servers, we are working with, only two were affected by the upgrade, so it's probably some specific circumstances or server misbehavior as you stated.

**Comment by terrafrost** [comment]
I tweaked the code for using the v2 algorithm suite somewhat:

```php
$methods = [
    'crypt' => array_intersect([
        'arcfour256',
        'arcfour128',
        'aes128-ctr',
        'aes192-ctr',
        'aes256-ctr',
        'twofish128-ctr',
        'twofish192-ctr',
        'twofish256-ctr',
        'aes128-cbc',
        'aes192-cbc',
        'aes256-cbc',
        'twofish128-cbc',
        'twofish192-cbc',
        'twofish256-cbc',
        'twofish-cbc',
        'blowfish-ctr',
        'blowfish-cbc',
        '3des-ctr',
        '3des-cbc'
    ], $ssh->getSupportedEncryptionAlgorithms()),
    'mac' => [
        'hmac-sha2-256',
        'hmac-sha1-96',
        'hmac-sha1',
        'hmac-md5-96',
        'hmac-md5'
    ],
    'comp' => ['none']
];

$ssh->setPreferredAlgorithms([
    'kex' => [
        'curve25519-sha256@libssh.org',
        'diffie-hellman-group-exchange-sha256',
        'diffie-hellman-group-exchange-sha1',
        'diffie-hellman-group14-sha1',
        'diffie-hellman-group14-sha256'
    ],
    'hostkey' => [
        'rsa-sha2-256',
        'rsa-sha2-512',
        'ssh-rsa',
        'ssh-dss'
    ],
    'client_to_server' => $methods,
    'server_to_client' => $methods
]);
```

Previously I was doing `diffie-hellman-group14-sha14` instead of `diffie-hellman-group14-sha256`. Also, I wasn't doing the array_intersect. That's necessary because there are some algorithms 3.0 won't even consider, depending on what extensions you have installed.
